### PR TITLE
feat(eventing): added error logs to the report and added retries for rest api requests

### PR DESCRIPTION
**Pull request by datacore-vvarakantham**
Added error logs during the report generation to report.
Removed the fields from the json report if there is any error while finding the field value.
Added retries to the mayastor rest api requests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.

### CODE REVIEW

File: call-home/src/bin/callhome/main.rs [hunk 1]
```diff
@@ -256,5 +294,81 @@ async fn generate_report(
...
+            "[{} {} {}:{}] {}",
+            log.timestamp, log.level, log.file, log.line, log.message
+        );
+        report.logs.push(log_string);
+    }
     report
 }
+
+// Define the LogsLayer
+struct LogsLayer {
```
Comment (niladrih) [inline call-home/src/bin/callhome/main.rs:310]: This bit and below definitely should move up to the lib level, than be at this bin level.

**Review by niladrih — COMMENTED** [review]
I dunno what all is going on here. But, changes look really cool!

File: call-home/src/bin/callhome/collector/report_models.rs [hunk 2]
```diff
@@ -488,20 +488,31 @@ impl Percentiles {
 #[derive(Serialize, Deserialize, Debug, Default)]
 #[serde(rename_all = "camelCase")]
 pub(crate) struct Report {
+    #[serde(skip_serializing_if = "String::is_empty")]
```
Comment (tiagolobocastro) [inline call-home/src/bin/callhome/collector/report_models.rs:491]: Is this even possible, shouldn't this be an error?

File: call-home/src/bin/callhome/collector/report_models.rs [hunk 3]
```diff
@@ -488,20 +488,31 @@ impl Percentiles {
...
     pub(crate) k8s_cluster_id: String,
-    pub(crate) k8s_node_count: u8,
+    #[serde(skip_serializing_if = "Option::is_none")]
```
Comment (tiagolobocastro) [inline call-home/src/bin/callhome/collector/report_models.rs:493]: I'd say these changes should be a separate commit

File: call-home/src/bin/callhome/main.rs [hunk 4]
```diff
@@ -172,58 +194,74 @@ async fn generate_report(
...
-            report.storage_node_count = nodes.len() as u8
+            report.storage_node_count = Some(nodes.len() as u8)
         }
         Err(err) => {
             error!("{:?}", err);
         }
     };
 
-    let pools = http_client.pools_api().get_pools(None).await;
+    let pools = Retry::spawn(retry_strategy.clone(), || async {
```
Comment (tiagolobocastro) [inline call-home/src/bin/callhome/main.rs:234]: I think this is not necessary, you can add a retry layer to the rest client itself through tower retry

File: call-home/src/bin/callhome/main.rs [hunk 5]
```diff
@@ -256,5 +294,81 @@ async fn generate_report(
...
+        let log_entry = LogEntry {
+            timestamp,
+            level: level.to_string(),
+            file,
+            line,
+            message: visitor.log,
+        };
+
+        // Only capture logs of level ERROR
+        if level == &Level::ERROR {
```
Comment (tiagolobocastro) [inline call-home/src/bin/callhome/main.rs:342]: I wonder if we also only want to capture logs with a certain field attribute. Example let's say one of the libraries starts generating a lot of error logs, this would generate a report with lots of data.

File: call-home/src/bin/callhome/main.rs [hunk 6]
```diff
@@ -256,5 +294,81 @@ async fn generate_report(
...
 
     report.nexus = Nexus::new(event_data);
+
+    let logs = logs.lock().unwrap();
+    for log in logs.iter() {
+        let log_string = format!(
+            "[{} {} {}:{}] {}",
+            log.timestamp, log.level, log.file, log.line, log.message
+        );
+        report.logs.push(log_string);
```
Comment (tiagolobocastro) [inline call-home/src/bin/callhome/main.rs:304]: Also consider here the size, it too large the report may fail to send? Should there be a cap?

**Comment by niladrih** [comment]
Closing the PR as these changes are included in #529.
